engine.lisp

```diff
@@ -148,9 +168,69 @@ Returns T if all constraints match, NIL otherwise."
   (loop for constraint in (cdr matcher)
         always (match-constraint constraint node)))
 
+(declaim (ftype (function (character list plump:node (function (plump:node) T))
+                          (values &optional null))
+                match-pair-depth))
+(defun match-pair-depth (combinator matcher parent extender)
+  "Match a combinator and matcher pair against a list of nodes.
+Returns a vector of matching nodes."
+  (declare (optimize speed))
+  (block match-pair-depth
+    (handler-case
+        (prog1 (values)
```

Comment:
There is no point in returning `(values)`, especially not when you think about performance. The 1-value return case is most likely to be optimised, anything else is bound to be slower. In general things like this are just bound to confuse, rather than aid in the readability. Please just get rid of this `(values)` trickery and let it return whatever happens to be returned.